awq/quantize/quantizer.py

```diff
@@ -542,6 +542,7 @@ def init_quant(self, n_samples=128, max_seq_len=512):
         best_device = get_best_device()
         modules[0] = modules[0].to(best_device)
         self.awq_model.move_embed(self.model, best_device)
+        self.awq_model.model.model.rotary_emb.to(best_device)
```

Comment:
Ah, I see what you mean now @SunMarc (you mentioned this over in the other PR). I'll go through each of the model files and check how this needs to be adjusted for each one. Sorry for not testing more fully: I tried this simple fix you proposed for several different models and it worked fine, but I see that it's not guaranteed to work depending on the model architecture.

Reply:
Thanks again @SunMarc! I checked (as in tried quantizing) all 26 of the models currently supported (including several variants where relevant) and fixed each of these separately only in cases where the fixes were needed, ensuring the code was adjusted to respect the model-specific `move_embed` method.

Note that a couple models are broken for various reasons, and I also noticed a somewhat tangential issue with llava and lava_next (seemingly related to this line in transformers for [llava](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L456) and [llava_next](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava_next/modeling_llava_next.py#L798) -- this can be tested with llava-hf/llava-1.5-7b-hf and llava-hf/llava-v1.6-mistral-7b-hf, respectively [as long as `device_map = None` is used).

### Notes

*(parentheses indicate models/variants tested)*

**fixed:**
- cohere (command r)
- gpt_neox (but model not currently not working due to a separate issue: `prev_op <class 'transformers.activations.FastGELUActivation'> not supported yet!`)
- llama (2, 3, 3.1)
- qwen2 (2, 2.5)
- stablelm (1; stablelm 2 not working for unrelated reasons)
- starcoder2

**no issues:**
- aquila
- baichuan
- gemma (1, 1.1)
- gemma2
- gpt_bigocde (bigcode/gpt_bigcode-santacoder)
- gptj
- internlm2
- minicpm
- mistral (0.1, 0.2, 0.3)
- mixtral
- opt
- phi3 (3, 3.5)
- qwen (1, 1.5)
- yi (1, 1.5)

**unfixed:**

current issue not noted but model not working (i.e., currently not working for unrelated reasons):
- bloom (see https://github.com/casper-hansen/AutoAWQ/issues/288)
deepseek_v2.py (deepseek-ai/DeepSeek-V2-Lite returns `AssertionError: ` [blank])
- falcon (see https://github.com/casper-hansen/AutoAWQ/issues/288)
- llava (1.5; seems to affected by a similar but slightly different issue: `Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument index in method wrapper_CUDA__index_select)`)
- llava-next (1.6)
- mpt (see https://github.com/casper-hansen/AutoAWQ/issues/293)

Reply:
@SunMarc: Turns out this issue is more complicated than I thought. I was testing this patch further, and unfortunately it does not work reliably as it can introduce a secondary error depending on the exact size of the model being quantized. For example, I didn't have issues with Qwen2.5 7B but 32B will fail about half-way through due to some issue with `accelerate` likely related to the `device_map`:

`NotImplementedError: Cannot copy out of meta tensor; no data!` (similar to the issue reported here on [StackOverflow](https://stackoverflow.com/questions/77547377/notimplementederror-cannot-copy-out-of-meta-tensor-no-data))

Long story short, I synced my PR (i.e., https://github.com/huggingface/transformers/pull/33742) in the transformers package to the latest code, reverted to the latest main branch code of AutoAWQ, and magically all issues are resolved (RoPE and these other tensor issues mentioned above). I'm honestly at a bit of a loss on what to do here.